lib/evaluate/traversal.h

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_TRAVERSAL_H_
+#define FORTRAN_EVALUATE_TRAVERSAL_H_
+
+#include "expression.h"
+#include <type_traits>
+
+// Implements an expression traversal utility framework.
```

Comment:
I think we need a comment explaining how to use this framework. That your class should extend `TraversalBase`. What it means to implement `Handle` vs. `Pre` and `Post` (and what happens if you implement all three). What happens after you call `Return`.

Reply:
Yes, of course.